perf(core): use `ngDevMode` to tree-shake `checkNoChanges`

arturovt opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
- [x] Refactoring (no functional changes, no api changes)


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

arturovt committed:
perf(core): use `ngDevMode` to tree-shake `checkNoChanges`

This commit adds `ngDevMode` guard to run `checkNoChanges` only
in dev mode (similar to how things work in other parts of Ivy runtime code).
The `ngDevMode` flag helps to tree-shake this code from production builds
(in dev mode everything will work as it works right now) to decrease production bundle size.

AndrewKushnir commented:
Hi @arturovt, thanks for creating the PRs to improve tree-shaking 👍 

I'd like to propose an idea to group all commits from other PRs (except the one that is already marked for merge) into a single PR (while keeping changes in separate commits). That should help review these changes as a whole and would simplify the process of landing these changes.

Thank you.

AndrewKushnir commented:
@arturovt another benefit is that you may need to update payload sizes for test/demo apps (reducing the size), so there might be code conflicts that you would need to resolve multiple times...

arturovt commented:
> @arturovt another benefit is that you may need to update payload sizes for test/demo apps (reducing the size), so there might be code conflicts that you would need to resolve multiple times...

Yeah that makes sense. I was just thinking that it's a bad idea to have changes, that affect different packages, in 1 PR.

arturovt committed:
perf(common): use `ngDevMode` to tree-shake warnings

This commit adds ngDevMode guard to show warnings only
in dev mode (similar to how things work in other parts of Ivy runtime code).
The ngDevMode flag helps to tree-shake these warnings from production builds
(in dev mode everything will work as it works right now) to decrease production bundle size.

arturovt committed:
perf(animations): use `ngDevMode` to tree-shake warning

This commit adds ngDevMode guard to show warning only
in dev mode (similar to how things work in other parts of Ivy runtime code).
The ngDevMode flag helps to tree-shake this warning from production builds
(in dev mode everything will work as it works right now) to decrease production bundle size.

arturovt committed:
perf(forms): use `ngDevMode` to tree-shake `_ngModelWarning`

This commit adds `ngDevMode` guard to call `_ngModelWarning` only
in dev mode (similar to how things work in other parts of Ivy runtime code).
The `ngDevMode` flag helps to tree-shake this function from production builds
(since it will act as no-op, in dev mode everything will work as it works right now)
to decrease production bundle size.

AndrewKushnir commented on packages/core/src/application_ref.ts:
```diff
@@ -764,7 +761,7 @@ export class ApplicationRef {
         view.detectChanges();
       }
-      if (this._enforceNoNewChanges) {
+      if (typeof ngDevMode === 'undefined' || ngDevMode) {
```
I think this one might be a bit tricky... For the cases when `ngDevMode` is not declared (i.e. for projects that use ViewEngine and don't use CLI), this code would not be tree-shaken and would be invoked in prod mode. This would be a change in behavior since currently it's not executed due to the `isDevMode()`-based check. I believe we may still achieve tree-shaking by just appending the `ngDevMode` to the existing check like this:
```suggestion
      if ((typeof ngDevMode === 'undefined' || ngDevMode) && isDevMode()) {
        for (let view of this._views) {
          view.checkNoChanges();
        }
      }
```
It'd be also helpful to leave a comment above the `if` statement to provide more context for future refactoring (when ViewEngine code is removed).

AndrewKushnir reviewed (changes requested):
Thanks for grouping commits into a single PR @arturovt 👍 

The changes look good, just left a comment were I believe we'd still need to use `isDevMode()` check. Could you please have a look when you get a chance?

arturovt committed:
fixup! perf(core): use `ngDevMode` to tree-shake `checkNoChanges`

arturovt committed:
fixup! perf(core): use `ngDevMode` to tree-shake `checkNoChanges`

arturovt committed:
fixup! perf(core): use `ngDevMode` to tree-shake `checkNoChanges`

arturovt committed:
fixup! perf(core): use `ngDevMode` to tree-shake `checkNoChanges`

AndrewKushnir reviewed (approved):
Thanks for addressing the feedback @arturovt 👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:345765972:BASE:345765992:1607123489519:d675307a).

mhevery reviewed (approved):
reviewed-for: global-approvers
